enhancements/deprecated-bundles.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Deprecating-Bundles
+authors:
+  - "@gallettilance"
+reviewers:
+  - "@ecordell"
+  - "@shawn-hurley"
+  - "@kevinrizza"
+approvers:
+  - "@ecordell"
+  - "@shawn-hurley"
+  - "@kevinrizza"
+creation-date: 2020-07-15
+last-updated: 2020-07-15
+status: implementable
+---
+
+# Deprecating Bundles
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+
+## Summary
+
+OPM will support marking versions as deprecated, truncating the update graph of a given package, and appending to the tail of channel. Deprecated versions will not be installable. If a deprecated version is installed prior to upgrade appropriate alerts will be fired on cluster. 
+
+## Motivation
+
+The default set of Operators provided on OCP clusters will soon (4.6+) be delivered through indexes and each OCP minor version will have a unique set of index images. Today, packages in one index will be duplicated to the next minor version index when it is available. This means operator owners releasing through the redhat, certified, or community pipelines will now need a way to specify which versions of their operator are incompatible or unsupported in the next minor version index.
+
+### Terminology
+
+#### Deprecated bundle
+
+A deprecated bundle is an upgradable but uninstallable bundle in the update graph of an operator.
+
+#### Terminated bundle
+
+A terminated bundle is neither upgradable nor installable.
+
+### Goals
+
+- A deprecated bundle in the update graph and any of its replacements (terminated bundles) will not be installed by OLM
```

Comment:
by "any of its replacements" do you mean "any of the versions it replaces"?

And this seems to imply that any version older than the deprecated version is implicitly considered `terminated` but it's not clear to me why that should be the case.  Can't both 1.2 and 1.3 of my operator be deprecated?

Reply:
If we say 1.2 is replaced by 1.3 in channel X and we deprecate 1.3, you would never be able to visit 1.3 from that channel.

Reply:
Offline:

Ben and I discussed this more. We need to make it more clear what happens based on some different channel upgrades. We also need to make it more clear how someone who is stranded (in the case above on 1.2) and now has no upgrade path how then can get out of that situation. 

My Opinion is that they can use a custom catalog to pull the old catalog and update it to a version that can be updated from, then switch back to the default catalog.

I imagine that we want to discuss this UX and the trade-offs.

Reply:
There's two cases in the "I'm on a not latest version and I'm going to be stranded" problem:

1. I'm on a version that actually needs to be stepped through before I can get to a supported version of my cluster upgrade.

In that case, you may actually need to do the upgrade to your operator before you execute the cluster upgrade.

2. I'm on a version that says it can't update directly to a supported version on the new cluster, but that's not really true.

In that case, either the package developer needs a way to update the graph to say "oh, you're actually good" OR the user just needs a way to force the upgrade anyway.

IMO we would eventually want a way to alert a user doing an upgrade that says "hey, you're going to be stranded, go do this update first" because of problem 1. For problem 2, that seems to be solved by giving package developers a way to overlay a new graph onto the index to allow that upgrade edge.

In either case though, I think we have some workarounds, so we can explore how to solve these problems more generally in the future.

Reply:
I too was confused that marking a version as deprecated implicitly marks all the versions it replaces as terminated. Just reading the enhancement, without this comment thread, I would have still been confused how a version ends up in the terminated state.

Reply:
Let's say I have OpenShift Serverless 1.7.0, 1.7.1, and 1.7.2 that are supported in OCP 4.5. They're also supported in OCP 4.6. Then, in OCP 4.7, I want to deprecate the Serverless 1.7.x versions. If I just mark the latest 1.7.x release as deprecated, anyone on a previous z-stream of 1.7 will now be terminated, not given an option to upgrade, and also not given an alert that they're on a deprecated version?

The realistic scenario I see is marking an entire major.minor version as deprecated instead of individual major.minor.patch versions. Is there a way to do that in this proposal? Or am I only marking a specific major.minor.patch version as deprecated and anyone not on exactly that patch version won't see a deprecation alert?

Reply:
@bbrowning keep in mind you're making it deprecated for a particular index image/ocp version.

So marking your 1.7.latest deprecated (thus terminating any old 1.7.x) will only impact consumers of the 4.7 index image.  Someone on ocp 4.6, using the 4.6 index image, could still upgrade to whatever 1.7.x versions were allowed previously (assuming you are distributing those in the 4.6 index image).

My expectation would be if you're making 1.7.latest as deprecated, that's because you expect people to move to 1.8.x, and that 1.8.x is available in the 4.7 index image.

So to play out the scenarios:

customer on 4.6:  they are not affected since you didn't impact the 4.6 index image.  They can upgrade to 1.7.latest.
customer on 4.7: they should already by on the 1.8.x version(emphasis on should be, but there no way to enforce that they are and they may have had to take explicit channel switching action to get to 1.8.x), but if they were not then yes, you've effectively stranded them at whatever 1.7.x version they were at, they have no way to upgrade to 1.7.latest or 1.8.x.

(i'm not clear on exactly how you specify that the deprecation only applies to a particular index image and not another one, i'll let @shawn-hurley explain how you actually define it)

Reply:
Discussed with Ben offline the outcomes were:

1. Whatever calls this deprecation needs to keep in mind that for the thing that is deprecated when it has new releases for the patch stream on the N version (previous ocp release) the N+1 needs to get the updated deprecation @htwaugh (this is not just before the release).

2. That we keep the UX of using the pipeline simple, so that user can declare the intent, and the pipeline can do the correct things to keep the index database up to date and correct.

Reply:
For (1), that's something the opm-wrapper we discussed would need to be aware of, right?